### PR TITLE
rec: add headerflags and ednsVersion as fields in protobuf messages

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

Please check the boundary change in `getQNameAndSubnet`  extra carefully.

This only adds the two new fields to the PB messages generated for client requests.

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [X] documented the code
- [X] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/recursordist/pdns_recursor.cc [hunk 1]
```diff
@@ -2027,14 +2027,19 @@ void getQNameAndSubnet(const std::string& question, DNSName* dnsname, uint16_t*
   const size_t headerSize = /* root */ 1 + sizeof(dnsrecordheader);
   const uint16_t arcount = ntohs(dhPointer->arcount);
 
-  for (uint16_t arpos = 0; arpos < arcount && questionLen > (pos + headerSize) && (lookForECS && !foundECS); arpos++) {
+  for (uint16_t arpos = 0; arpos < arcount && questionLen >= (pos + headerSize) && (lookForECS && !foundECS); arpos++) {
```
Comment (rgacogne) [inline pdns/recursordist/pdns_recursor.cc:2030]: Right now we are only executing this loop if we are looking for ECS, and we know there is not going to be an ECS option if `questionLen == (pos + headerSize)`. I guess the question is, do we want to make to always enable this parsing so we can get the EDNS version? If so we need to get rid of the `lookForECS` condition, and likely change the one on line 2259.

File: regression-tests.recursor-dnssec/test_Protobuf.py [hunk 2]
```diff
@@ -342,6 +349,8 @@ def testA(self):
         # check the protobuf messages corresponding to the UDP query and answer
         msg = self.getFirstProtobufMessage()
         self.checkProtobufQuery(msg, dnsmessage_pb2.PBDNSMessage.UDP, query, dns.rdataclass.IN, dns.rdatatype.A, name)
+        # wire format, RD and CD set in headerflags, plus DO bit in flags part of EDNS Version
+        self.checkProtobufHeaderFlagsAndEDNSVersion(msg, 0x0110, 0x00008000)
```
Comment (rgacogne) [inline regression-tests.recursor-dnssec/test_Protobuf.py:353]: We might want to check a packet-cache hit as well?

File: pdns/dnsmessage.proto [hunk 3]
```diff
@@ -2,19 +2,19 @@
...
  * in the Software without restriction, including without limitation the rights
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  * copies of the Software, and to permit persons to whom the Software is
  * furnished to do so, subject to the following conditions:
- *
+ * 
  * The above copyright notice and this permission notice shall be included in all
  * copies or substantial portions of the Software.
- *
+ *
```
[the remark this answers is not in the document]
Comment (omoerbeek) [reply pdns/dnsmessage.proto:17]: I submitted a PR to the dnsmessage repo.

File: pdns/dnsmessage.proto [hunk 4]
```diff
@@ -188,6 +188,8 @@ message PBDNSMessage {
   optional uint64 workerId = 25;                // Thread id
   optional bool packetCacheHit = 26;            // Was it a packet cache hit?
   optional uint32 outgoingQueries = 27;         // Number of outgoing queries used to answer the query
+  optional uint32 headerFlags = 28;             // Flags field in wire format
```
[the remark this answers is not in the document]
Comment (omoerbeek) [reply pdns/dnsmessage.proto:191]: Same

File: pdns/recursordist/pdns_recursor.cc (hunk 1 above)
Comment (omoerbeek) [reply pdns/recursordist/pdns_recursor.cc:2030]: If we're PB logging, `needECS` is set to true (2220) and `getQNameAndSubnet` is called with a set `ednssubnet`. I think I'll rename `needECS` to `needEDNSparse` and adapt `getQNameAndSubnet`.

Comment (rgacogne) [reply pdns/recursordist/pdns_recursor.cc:2030]: Ah, I was looking for something like that and I still missed it, thanks!

File: pdns/recursordist/pdns_recursor.cc [hunk 5]
```diff
@@ -2027,22 +2026,27 @@ void getQNameAndSubnet(const std::string& question, DNSName* dnsname, uint16_t*
...
     if (question.at(pos) != 0) {
       /* not an OPT, bye. */
       return;
     }
 
     pos += 1;
     const auto* drh = reinterpret_cast<const dnsrecordheader*>(&question.at(pos)); // NOLINT(cppcoreguidelines-pro-type-reinterpret-cast)
+    if (ntohs(drh->d_type) == QType::OPT) {
+      uint32_t edns{};
+      memcpy(&edns, &drh->d_ttl, sizeof(edns)); // dh is not neccesarily aligned, so no uint32 assignment can be done
```
Comment (rgacogne) [inline pdns/recursordist/pdns_recursor.cc:2039]:
```suggestion
      memcpy(&edns, &drh->d_ttl, sizeof(edns)); // drh is not neccesarily aligned, so no uint32 assignment can be done
```

**Review by rgacogne — APPROVED** [review]

**Comment by omoerbeek** [comment]
@neilcook Are you happy with this PR?

**Review by neilcook — APPROVED** [review]

**Pull request merged**